File: Alumbrado/alumbrado.py
import math
import sys
from collections import defaultdict, deque
from itertools import *
from functools import *

input = sys.stdin.readline
write = sys.stdout.write
def fast_print(x): write(str(x) + '\n')

class pt:
    def __init__(self, x=0, y=0):
        self.x = x
        self.y = y

    def __add__(self, p): return pt(self.x + p.x, self.y + p.y)
    def __sub__(self, p): return pt(self.x - p.x, self.y - p.y)
    def __mul__(self, other):
        if isinstance(other, pt): return self.x * other.x + self.y * other.y
        return pt(self.x * other, self.y * other)   
    def cross(self, p): return self.x * p.y - self.y * p.x
    def __lt__(self, p): return (self.x, self.y) < (p.x, p.y)
    def __eq__(self, p): return self.x == p.x and self.y == p.y
    def __repr__(self): return f"pt({self.x}, {self.y})"

def sq(x): return x * x
def dist2(p, q): return sq(p.x - q.x) + sq(p.y - q.y)
def sarea2(p, q, r): return (q - p).cross(r - q)
def col(p, q, r): return sarea2(p, q, r) == 0
def ccw(p, q, r): return sarea2(p, q, r) > 0
def quad(p): return (int(p.x < 0)) ^ (3 * int(p.y < 0))
def compare_angle(p, q):
    if quad(p) != quad(q): return quad(p) < quad(q)
    return ccw(q, pt(0, 0), p)

def isinseg(p, r): 
    a, b = r[0] - p, r[1] - p
    return a.cross(b) == 0 and a * b <= 0

def interseg(r, s): 
    if (isinseg(r[0], s) or isinseg(r[1], s) or isinseg(s[0], r) or isinseg(s[1], r)):
        return True
    return (ccw(r[0], r[1], s[0]) != ccw(r[0], r[1], s[1]) and ccw(s[0], s[1], r[0]) != ccw(s[0], s[1], r[1]))
def read_pt():
    x, y = map(int, input().split()) 
    return pt(x, y)
    
def solve():
    a,s,d,f = map(int, input().split())
    A, B = pt(a,s), pt(d,f)
    Q = read_pt() 
    P = read_pt()
    ff = (A, B)
    ss = (Q, P)
    fast_print("NO" if interseg(ff, ss) else "SI")

def main():
    t = int(input())
    for _ in range(t):
        solve()

if __name__ == '__main__':
    main()

File: Alumbrado/alumbrado.java
import java.util.*;
import java.io.*;

public class alumbrado {
    static long sq(long x) { return x * x; }
    static long[] pt(long x, long y) { return new long[]{x, y}; }
    static long[] add(long[] p, long[] q) { return pt(p[0]+q[0], p[1]+q[1]); }
    static long[] sub(long[] p, long[] q) { return pt(p[0]-q[0], p[1]-q[1]); }
    static long[] mul(long[] p, long  c)  { return pt(p[0]*c,    p[1]*c);    }
    static long dot  (long[] p, long[] q) { return p[0]*q[0] + p[1]*q[1]; }
    static long cross(long[] p, long[] q) { return p[0]*q[1] - p[1]*q[0]; }
    static long dist2(long[] p, long[] q) { return sq(p[0]-q[0]) + sq(p[1]-q[1]); }
    static long sarea2(long[] p, long[] q, long[] r) { return cross(sub(q, p), sub(r, q));}
    static boolean col(long[] p, long[] q, long[] r) { return sarea2(p,q,r) == 0; }
    static boolean ccw(long[] p, long[] q, long[] r) { return sarea2(p,q,r) >  0; }
    static boolean isinseg(long[] p, long[] a, long[] b) {
        long[] u = sub(a, p), v = sub(b, p);
        return cross(u, v) == 0 && dot(u, v) <= 0;
    }

    
    static boolean interseg(long[] r0, long[] r1, long[] s0, long[] s1) {
        if (isinseg(r0,s0,s1) || isinseg(r1,s0,s1) || isinseg(s0,r0,r1) || isinseg(s1,r0,r1)) return true;
        return ccw(r0,r1,s0) != ccw(r0,r1,s1) && ccw(s0,s1,r0) != ccw(s0,s1,r1);
    }

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();
    static String next() throws IOException {
        while (st == null || !st.hasMoreTokens()) st = new StringTokenizer(br.readLine());
        return st.nextToken();
    }

    static long nextLong() throws IOException { return Long.parseLong(next()); }
    static long[] readPt() throws IOException { return pt(nextLong(), nextLong()); }

    
    static void solve() throws IOException {
        long[] A = readPt(), B = readPt();
        long[] Q = readPt(), P = readPt();

        sb.append(interseg(A, B, Q, P) ? "NO" : "SI").append('\n');
    }

    public static void main(String[] args) throws IOException {
        int t = (int) nextLong();
        while (t-- > 0) solve();
        System.out.print(sb);
    }
}